tcp/client.go

```diff
@@ -5,32 +5,83 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"strings"
 
 	"github.com/influxdb/influxdb/cluster"
 	"github.com/influxdb/influxdb/tsdb"
 )
 
+const maxConnections = 500
+
+var errMaxConnectionsExceeded = fmt.Errorf("can not exceed max connections of %d", maxConnections)
+
+type clientConn struct {
+	client *Client
+	addr   string
+}
+
+func newClientConn(addr string, c *Client) *clientConn {
+	return &clientConn{
+		addr:   addr,
+		client: c,
+	}
+}
+func (c *clientConn) dial() (net.Conn, error) {
+	if c.client.poolSize() > maxConnections {
+		return nil, errMaxConnectionsExceeded
+	}
+
+	conn, err := net.Dial("tcp", c.addr)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
 type Client struct {
-	conn net.Conn
+	pool *connectionPool
 }
 
 func NewClient() *Client {
-	return &Client{}
+	return &Client{
+		pool: newConnectionPool(),
+	}
+}
+
+func (c *Client) poolSize() int {
+	if c.pool == nil {
+		return 0
+	}
+
+	return c.pool.size()
+}
+
+func (c *Client) dial(addr string) (net.Conn, error) {
+	addr = strings.ToLower(addr)
+	// if we don't have a connection pool for that addr yet, create one
+	_, ok := c.pool.getPool(addr)
+	if !ok {
+		conn := newClientConn(addr, c)
+		p, err := NewChannelPool(1, 3, conn.dial)
+		if err != nil {
+			return nil, err
+		}
+		c.pool.setPool(addr, p)
+	}
+	return c.pool.conn(addr)
 }
 
-func (c *Client) Dial(addr string) error {
-	conn, err := net.Dial("tcp", addr)
+func (c *Client) WriteShard(addr string, shardID uint64, points []tsdb.Point) error {
```

Comment:
I would't expect `WriteShard` to take a connection address. I would expect a `Client` to be tied to a specific address already and `WriteShard` to simply take the `shardID` and the `points`.